sdk/metricsadvisor/ai-metrics-advisor/review/ai-metrics-advisor.api.md

```diff
@@ -0,0 +1,1152 @@
+## API Report File for "@azure/ai-metrics-advisor"
+
+> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).
+
+```ts
+
+import * as coreHttp from '@azure/core-http';
+import { OperationOptions } from '@azure/core-http';
+import { PagedAsyncIterableIterator } from '@azure/core-paging';
+import { PipelineOptions } from '@azure/core-http';
+import { RestResponse } from '@azure/core-http';
+
+// @public
+export interface Alert {
+    createdOn?: Date;
+    id: string;
+    modifiedOn?: Date;
+    timestamp?: Date;
+}
+
+// @public (undocumented)
+export interface AlertSnoozeCondition {
+    autoSnooze: number;
+    onlyForSuccessive: boolean;
+    snoozeScope: SnoozeScope;
+}
+
+// @public
+export interface Anomaly {
+    createdOn?: Date;
+    detectionConfigurationId: string;
+    dimension: Record<string, string>;
+    metricId?: string;
+    modifiedOn?: Date;
+    severity: "Low" | "Medium" | "High";
+    status?: "Active" | "Resolved";
+    timestamp: Date;
+}
+
+// @public
+export interface AnomalyAlertConfiguration {
+    crossMetricsOperator?: AnomalyAlertingConfigurationCrossMetricsOperator;
+    description?: string;
+    hookIds: string[];
+    id: string;
+    metricAlertConfigurations: MetricAlertConfiguration[];
+    name: string;
+}
+
+// @public
+export interface AnomalyAlertConfigurationPatch {
+    crossMetricsOperator?: AnomalyAlertingConfigurationCrossMetricsOperator;
+    description?: string;
+    hookIds?: string[];
+    metricAlertConfigurations?: MetricAlertConfiguration[];
+    name?: string;
+}
+
+// @public
+export type AnomalyAlertingConfigurationCrossMetricsOperator = "AND" | "OR" | "XOR";
+
+// @public
+export interface AnomalyDetectionConfiguration {
+    description?: string;
+    id: string;
+    metricId: string;
+    name: string;
+    seriesDetectionConditions?: MetricSingleSeriesDetectionCondition[];
+    seriesGroupDetectionConditions?: MetricSeriesGroupDetectionCondition[];
+    wholeSeriesDetectionCondition: MetricDetectionCondition;
+}
+
+// @public
+export interface AnomalyDetectionConfigurationPatch {
+    description?: string;
+    name?: string;
+    seriesDetectionConditions?: MetricSingleSeriesDetectionCondition[];
+    seriesGroupDetectionConditions?: MetricSeriesGroupDetectionCondition[];
+    wholeSeriesDetectionCondition?: MetricDetectionCondition;
+}
+
+// @public
+export type AnomalyDetectorDirection = "Both" | "Down" | "Up";
+
+// @public
+export type AnomalyValue = "AutoDetect" | "Anomaly" | "NotAnomaly";
+
+// @public
+export type AzureApplicationInsightsDataFeedSource = {
+    dataSourceType: "AzureApplicationInsights";
+    dataSourceParameter: AzureApplicationInsightsParameter;
+};
+
+// @public
+export type AzureApplicationInsightsDataFeedSourcePatch = {
+    dataSourceType: "AzureApplicationInsights";
+    dataSourceParameter?: AzureApplicationInsightsParameter;
```

Comment:
```typescript
export type DataFeedSourcePatch = Omit<DataFeedSource, "dataSourceParameter"> &
  { [P in "dataSourceParameter"]?: DataFeedSource[P] };
```
This seems equivalent.  One concern I had about advanced TypeScript type is that consumers don't always get good help from IDE/IntelliSense.

Reply:
I would like to avoid declaring the patch types if possible, like this:

```ts
interface Foo {
    kind: "foo";
    dataSource: string;
}

type OptionalProp<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;

function test(foo: OptionalProp<Foo, "dataSource">) {
    // foo.dataSource is optional now
}
```

Reply:
`DataFeedSourcePatch` is part of `DataFeedPatch`, and I use `DataFeedSourcePatch.dataSourceType` as the discriminator so I am making it required. There's also a service bug which requires data source type when updating.

```typescript
export interface DataFeedPatch {
  name?: string;
  source: DataFeedSourcePatch;
  // other optional properties
}
```

Also not all of data feed properties can be updated.
```typescript
updateDataFeed(id: string, patch: Omit<DataFeed, "source" | "id" | "metricIds" | ...> & { source: OptionalProp<DataFeedSource, "dataSourceParameter"> }, ...)
```
That's when I feel an explicit patch type is better.